User Guide for Federated Services.

quinton-hoole opened the pull request:
cc: @mml @madhusudancs @davidopp for optional technical and stylistic review.

@johndmulhausen For docs review.

@colhom, @mfanjie @kelseyhightower FYI and comments welcome.

quinton-hoole commented:
@mml @madhusudancs Ping for tech review.
@johndmulhausen Ping for docs review.  I'd very much like to get this merged today/as there is some publicity going out on Thursday (14th), and this needs to merge before then.

mml commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,234 @@
+```
+
+```
+kubectl --context=gce-asia-east1a get svc nginx
````
May as well use the same syntax (`services`) as above.

madhusudancs replied:
Yeah. +1.

mml commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,234 @@
+You can verify this by checking in each of the underlying clusters, for example:
+
+```
+kubectl --context=gce-asia-east1-a get services nginx
````
These examples need a namespace.

quinton-hoole replied:
I think they work fine with no namespace, or the namespace configured in the context. I'd rather not clutter the example with the --namespace argument given that it's largely irrelevant?  OK with that?

mml commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,234 @@
+kubectl --context=gce-asia-east1-a get services nginx
+```
+
+```
````
If you use

```
```console
```

then command and output get formatted differently:

```console
$ kubectl blah blah
NAME
blah
```

But this is a style issue.  Not sure if we have established a style on this sort of thing.

quinton-hoole replied:
I'll leave it to the style reviewers to decide/standardize.  Leaving as is for now.

mml commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,234 @@
+No events.
+```
+
+Note the the 'LoadBalancer Ingress' addresses of your Federated Service correspond with the 'LoadBalancer Ingress' addresses of all of the underlying Kubernetes services (once these have been allocated - this may take a few seconds).  For inter-cluster and inter-cloud-provider networking between service shards to work correctly, your services need to have an externally visible IP address.  Servic ...
````
"the the"

quinton-hoole replied:
Good catch, done.

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+
+Note the the 'LoadBalancer Ingress' addresses of your Federated Service correspond with the 'LoadBalancer Ingress' addresses of all of the underlying Kubernetes services (once these have been allocated - this may take a few seconds).  For inter-cluster and inter-cloud-provider networking between service shards to work correctly, your services need to have an externally visible IP address.  Servic ...
+
+Note also what we have not yet provisioned any backend Pods to receive the network traffic directed to these addresses (i.e. 'Service Endpoints'), so the Federated Service does not yet consider these to be healthy service shards, and has accordingly not yet added their addresses to the DNS records for this Federated Service (more on this aspect later). 
```
s/what we/that we/

quinton-hoole replied:
Yup, done.

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+                        us-central1-f us-central1-a us-central1-b us-central1-c \
+						us-east1-d us-east1-c us-east1-b
+do
+  kubectl --context=$CLUSTER run nginx --image=nginx:1.11.1-alpine --port=80
```
I guess `kubectl run` automatically adds the `run=nginx` label?  If so, that's nice and ties the example together, but you might wish to say so explicitly.

quinton-hoole replied:
Yes, that's it.  Will add it.

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+
+## Verifying Public DNS Records
+
+Once the above Pods have successfully started and have begun listening for connections, Kubernetes in each cluster (via automatic health checks) will report them as healthy endpoints of the service in that cluster.  The Cluster Federation will in turn consider each of these service 'shards' to be healthy, and place them in serving by automatically configuring corresponding public DNS records.  Yo ...
```
More style: we should decide if shard is a term we're going to use and then decide on whether to scare quote it or not.

We need better terminology than what we've got right now, which is often cumbersomely throwing around modifiers like "underlying", "within the federation", or "in each cluster".

quinton-hoole replied:
Yes, I agree 100%.  I'm in favor of clearly defining the term 'shard' up front, and then using it consistently.  Does that sound good?

madhusudancs commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+---
+# Kubernetes Federated Services and Cross-Cluster Service Discovery
+
+This guide explains how to use Kubernetes Federated Services to deploy a common Service across multiple Kubernetes clusters to achieve cross-cluster service discovery and availibility zone fault tolerance for your Kubernetes applications.
```
I think this sentence is too long to comprehend. I lost track at "common Service across multiple Kubernetes clusters". That might be also be a good checkpoint to break into a new sentence.

quinton-hoole replied:
Agreed, will do.

mml commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,234 @@
+$aws route53 list-resource-record-sets --hosted-zone-id Z3ECL0L9QLOVBX
+```
+
+Whatever DNS provider you use, any DNS query tool (for example 'dig' or 'nslookup') will of course also allow you to see the records created by the Federation for you.
````
...once the assorted TTLs have expired.

quinton-hoole replied:
Actually not, if you point them straight at your DNS provider, but yes, worth adding a note to prevent people running foul of TTL's

madhusudancs commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+
+## Hybrid Cloud Capabilities
+
+Federations of Kubernetes Clusters can include clusters running in different cloud providers (e.g. Google Cloud, AWS), and on-premise (e.g. on OpenStack).  Simply create all of the clusters that you require, in the appropriate cloud providers and/or locations, and register each cluster's API endpoint and credentials with your Federation API Server  (See the [federation admin guide](/docs/admin/fe ...
```
> and on-premise (e.g. on OpenStack).

on-premises ...

quinton-hoole replied:
Yup, good point.

madhusudancs commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+
+Federations of Kubernetes Clusters can include clusters running in different cloud providers (e.g. Google Cloud, AWS), and on-premise (e.g. on OpenStack).  Simply create all of the clusters that you require, in the appropriate cloud providers and/or locations, and register each cluster's API endpoint and credentials with your Federation API Server  (See the [federation admin guide](/docs/admin/fe ...
+
+Thereafter, your applications and services can span different clusters and clout providers as described in more detail below.
```
> and clout providers 

cloud ...

mml replied:
This is my favorite typo so far this year.

quinton-hoole replied:
I noticed it earlier, but intentionally left it in for entertainment value :-)

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+
+## Handling Failures of Backend Pods and Whole Clusters
+
+Standard Kubernetes service cluster-IP's already ensure that non-responsive individual Pod endpoints are automatically taken out of service with low latency (a few seconds). In addition, as aluded above, the Kubernetes Cluster Federation system automatically monitors the health of clusters and the endpoints behind all of the shards of your Federated Service, taking shards in and out of service as ...
```
*alluded

madhusudancs replied:
> alternative cluster in in the case

deduplicate "in"s

madhusudancs replied:
Also, it might be useful to enable automatic line wrapping for markdown in your editor. Long paragraphs are too difficult to review.

quinton-hoole replied:
Yup, agreed, done.

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+#### I cannot connect to my Cluster Federation API
+Check that your
+
+1. Client (typically kubectl) is correctly configured (including API endpoints and login credentials), and
```
Do you see a different errmsg if the credentials are wrong?  We could spell that out if so.

quinton-hoole replied:
Agreed, will flesh out further in a separate PR.

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+Check that:
+
+1. Your clusters are correctly registered in the Cluster Federation API (`kubectl describe clusters`)
+2. Your clusters are all 'Active'.  This means that the cluster Federation system was able to connect and authenticate against the clusters' endpoints.  If not, consult the logs of the federation-controller-manager pod to ascertain what the failure might be. (`kubectl logs federation-controller-manager --namespace federation`)
```
The pod isn't actually named `federation-controller-manager`.

`kubectl --namespace=federation logs $(kubectl get pods --namespace=federation -l module=federation-controller-manager -oname)`

quinton-hoole replied:
Thanks, done.  That was from memory :-)  which is obviously faulty.

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+
+1. Your clusters are correctly registered in the Cluster Federation API (`kubectl describe clusters`)
+2. Your clusters are all 'Active'.  This means that the cluster Federation system was able to connect and authenticate against the clusters' endpoints.  If not, consult the logs of the federation-controller-manager pod to ascertain what the failure might be. (`kubectl logs federation-controller-manager --namespace federation`)
+3. That the login credentials provided to the Cluster Federation API for the clusters have the correct authorization and quota to create services in the relevant namespace in the clusters.  Again you should see associated error messages providing more detail in the above log file if this is not the case.
```
We should file a bug to issue events for this so that `kubectl --context=federation describe {svc,cluster}` will show you the problem.

quinton-hoole replied:
Agreed. Will do.

mml commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+Check that:
+
+1. The DNS registrar that manages your federation DNS domain has been correctly configured to point to your configured DNS provider's nameservers.  See for example [Google Domains Documentation](https://support.google.com/domains/answer/3290309?hl=en&ref_topic=3251230) and [Google Cloud DNS Documentation](https://cloud.google.com/dns/update-name-servers), or equivalent guidance from your domain r ...
+
```
Perhaps add "This troubleshooting guide did not help" and point them to the support channels, particularly mentioning slack?

quinton-hoole replied:
Good idea.  Done.

mml commented:
Very thorough and clear.  I think.  I'm a little too close to be sure.  :-)

madhusudancs commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,234 @@
+
+With the introduction of Federated Services and Cross-Cluster Service Discovery, this concept is extended to cover Kubernetes services running in any other cluster across your Cluster Federation, globally.  To take advantage of this extended range, you use a slightly different DNS name (of the form "<servicename>.<namespace>.<federationname>", e.g. myservice.mynamespace.myfederation) to resolve F ...
+
+So, using our NGINX example service above, and the Federated Service DNS name form just described, let's consider an example: A Pod in a cluster in the `us-central1-f` availability zone needs to contact our NGINX service.  Rather than use the service's traditional cluster-local DNS name (```"nginx.mynamespace"```, which is automatically expanded to ```"nginx.mynamespace.svc.cluster.local"```) it  ...
````
We actually return a `CNAME` now and `KubeDNS` also appends the resolved `A` record to the response, but it is still worth documenting that behavior in this user guide.

madhusudancs commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,234 @@
+
+So, using our NGINX example service above, and the Federated Service DNS name form just described, let's consider an example: A Pod in a cluster in the `us-central1-f` availability zone needs to contact our NGINX service.  Rather than use the service's traditional cluster-local DNS name (```"nginx.mynamespace"```, which is automatically expanded to ```"nginx.mynamespace.svc.cluster.local"```) it  ...
+
+But if the service does not exist in the local cluster (or it exists but has no healthy backend pods), the DNS query is automatically expanded to ```"nginx.mynamespace.myfederation.svc.us-central1-f.example.com"`` (i.e. logically "find the external IP of one of the shards closest to my availability zone").  This expansion is performed automatically by KubeDNS, which returns the associated CNAME r ...
````
> "nginx.mynamespace.myfederation.svc.us-central1-f.example.com"

Three closing back ticks instead of two here. Causing rendering problems right now.

madhusudancs replied:
> This results in automitic

automatic ...

madhusudancs commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,234 @@
+1. Your federation name, DNS provider, DNS domain name are configured correctly.  Consult the [federation admin guide](/docs/admin/federation/) or  [tutorial](https://github.com/kelseyhightower/kubernetes-cluster-federation) to learn
+how to configure your Cluster Federation system's DNS provider (or have your cluster administrator do this for you).
+2. Confirm that the Cluster Federation's service-controller is successfully connecting to and authenticating against your selected DNS provider (look for `service-controller` errors or successes in the output of `kubectl logs federation-controller-manager --namespace federation`)
+3. Confirm that the Cluster Federation's service-controller is successfully creating DNS records in your DNS provider (or outputting errors in it's logs explaining in more detail what's failing).
```
Adding the `kubectl logs` command also here is useful.

quinton-hoole replied:
Agreed, will add in followup PR, as I want to confirm all of the sample debugging commands in different environments.

madhusudancs commented:
Overall, it looks good to me. A few comments before actually adding the label.

A more general comment.

At some point you talk about service shard adoption, but it might also be worth noting what happens when you delete a federated service, particularly to the adopted shards.

quinton-hoole commented:
Thanks for the review guys.  I've addressed most of the comment, and will do the remainder in a followup PR.

quinton-hoole commented:
@johndmulhausen Docs review please.  Fairly urgent now, as it ideally needs to merge today/tomorrow.

johndmulhausen commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,378 @@
+---
+---
+# Kubernetes Federated Services and Cross-Cluster Service Discovery
```
Remove; the title of the document should be in /_data/guides.yml, where this page must be added in order to show up in the left-navigation.

quinton-hoole replied:
Done

johndmulhausen commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,378 @@
+cluster if it exists and is healthy, or the closest healthy shard in a
+different cluster if it does not.
+
+## Hybrid Cloud Capabilities
```
Sentence casing for headings, not title casing (which is only for page titles). Thus:

`## Hybrid cloud capabilities`

This goes for the entire PR, only gonna mention it once here.

quinton-hoole replied:
Done.

johndmulhausen commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,378 @@
+
+This is done in the usual way, for example:
+
+```
````
```shell

quinton-hoole replied:
Done.

johndmulhausen commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,378 @@
+underlying Kubernetes services (once these have been allocated - this
+may take a few seconds). For inter-cluster and inter-cloud-provider
+networking between service shards to work correctly, your services
+need to have an externally visible IP address. Service Type:
```
I would delete the last sentence; it kind of prescribes a certain use case and makes it seem like there aren't other options.

quinton-hoole replied:
Added a note regarding other alternatives, rather than removing this guidance.

johndmulhausen commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,378 @@
+  kubectl --context=$CLUSTER run nginx --image=nginx:1.11.1-alpine --port=80
+done
+```
+Note that `kubectl run` automatically adds the `run=nginx` labels required to associate the backend pods with their services.
````
Always leave the lines above and below code example blocks blank.

johndmulhausen commented on docs/user-guide/federation/federated-services.md:
```diff
@@ -0,0 +1,378 @@
+Once the above Pods have successfully started and have begun listening
+for connections, Kubernetes in each cluster (via automatic health
+checks) will report them as healthy endpoints of the service in that
+cluster. The Cluster Federation will in turn consider each of these
```
...Kubernetes will report them as healthy endpoints of the service in that cluster.

quinton-hoole commented on docs/user-guide/federation/federated-services.md:
````diff
@@ -0,0 +1,378 @@
+```
+kubectl --context=federation-cluster describe services nginx
+```
+
````
Done.

quinton-hoole commented:
Thanks @johndmulhausen !  I've addressed all your comments, and added a cross references from /docs/user-guide/connecting-applications.md

PTAL. I think that this is ready for merge now.

quinton-hoole commented:
@johndmulhausen Ping.  Can I merge this please?
